Change Loader to Type[BaseConstructor] in yaml.add_constructor

Stannislav opened the pull request:
This fixes a wrong type in the PyYAML stubs.

The last parameter in `yaml.add_constructur(tag, constructor, Loader)` is the loader class and not its instance. This is consistent with both the [PyYAML docs](https://pyyaml.org/wiki/PyYAMLDocumentation) and the way it's implemented in the [PyYAML package](https://github.com/yaml/pyyaml/blob/ee37f4653c08fc07aecff69cfd92848e6b1a540e/lib/yaml/__init__.py#L342).

In fact, since the only requirement on the `Loader` parameter is to have the `Loader.add_constructor` class method, which is defined in the `BaseConstructor` base class, we should use the corresponding base class type `Type[BaseConstructor]`

Simlar types can be added to the similar functions in the same block:
- `def add_implicit_resolver(...)`
- `def add_path_resolver(...)`
- `def add_constructor(...)`
- `def add_multi_constructor(...)`
- `def add_representer(...)`
- `def add_multi_representer(...)`

JelleZijlstra commented:
Thank you! Feel free to change the other instances of this issue too.

Stannislav commented:
Hi @JelleZijlstra, thanks for having a look. I just added annotations to the other functions mentioned in the description.

It would be good if someone more experience than me in both typing and the yaml-library could glance over it, just to see if all makes sense.

Also, I'm not 100% sure if writing something like `Type[Any]` in `add_path_resolver` makes sense.

Akuli commented on stubs/PyYAML/yaml/__init__.pyi:
```diff
@@ -258,12 +261,32 @@ def safe_dump(
+    Loader: Optional[Type[BaseResolver]] = ...,
+    Dumper: Type[BaseResolver] = ...,
+) -> None: ...
+def add_constructor(
+    tag: _Str, constructor: Callable[[_Constructor, Node], Any], Loader: Optional[Type[_Constructor]] = ...
+) -> None: ...
```
If `Loader` is not specified (or is specified as `None`), mypy can't figure out what type `_Constructor` should be. For example, this...

```python
from yaml import add_constructor
add_constructor("lol", (lambda x, y: "wut"))
```

...fails with this error:

```
(env) akuli@akuli-desktop:~/typeshed$ MYPYPATH=stubs/PyYAML mypy foo.py
foo.py:2: error: Cannot infer type argument 1 of "add_constructor"
Found 1 error in 1 file (checked 1 source file)
```

If I read the source code correctly, in this case it can be `Loader`, `FullLoader` or `UnsafeLoader`. To do this, you need two overloads: one when `Loader` is not given or it is None, and one when it is given.

```suggestion
@overload
def add_constructor(
    tag: _Str, constructor: Callable[[Loader | FullLoader | UnsafeLoader, Node], Any], Loader: None = ...
) -> None: ...
@overload
def add_constructor(
    tag: _Str, constructor: Callable[[_Constructor, Node], Any], Loader: Type[_Constructor]
) -> None: ...
```

The same applies to other functions that use `_Constructor` or `_Representer` too.

Stannislav replied:
Thanks for the explanation! I addressed them in 8f56493 (+ a small fix in adc766b).

In `add_representer` it's actually a peculiar case that the default of the last parameter is not `None` but `Dumper`, which itself is a subclass of `BaseRepresenter`. So it seems that leaving out this last argument in the first overload is fine:
```python
@overload
def add_representer(data_type: Type[_T], representer: Callable[[Dumper, _T], Node]) -> None: ...
@overload
def add_representer(
    data_type: Type[_T], representer: Callable[[_Representer, _T], Node], Dumper: Type[_Representer] = ...
) -> None: ...
```
because calling `add_representer` with its default third parameter passed explicitly is actually covered by the second overload. (This didn't work for `add_constructor` since the default is `None`, hence the fix in adc766b)

I hope it's fine that way.

Akuli commented:
We changed all stubs to use the new `foo | None` syntax instead of `Optional[foo]` in #5872, and that conflicts with this PR. To fix it, pull from our master and then resolve the conflicts:

```
$ git pull https://github.com/python/typeshed
```

Please also use the newer `foo | None` syntax in this PR.

>Also, I'm not 100% sure if writing something like Type[Any] in add_path_resolver makes sense.

It makes sense. PyYAML checks whether it is `str` or `list` or some other types.

Stannislav commented:
> We changed all stubs to use the new `foo | None` syntax instead of `Optional[foo]` in #5872, and that conflicts with this PR. To fix it, pull from our master and then resolve the conflicts:
> 
> ```
> $ git pull https://github.com/python/typeshed
> ```
> 
> Please also use the newer `foo | None` syntax in this PR.
> 
> > Also, I'm not 100% sure if writing something like Type[Any] in add_path_resolver makes sense.
> 
> It makes sense. PyYAML checks whether it is `str` or `list` or some other types.

Hi @Akuli, thanks for reviewing. Merged master in bc77aed and also addressed your comment in the [thread above](https://github.com/python/typeshed/pull/5828#discussion_r684747087).

Akuli commented on stubs/PyYAML/yaml/__init__.pyi:
```diff
@@ -258,12 +261,46 @@ def safe_dump(
+    tag: _Str, constructor: Callable[[Loader | FullLoader | UnsafeLoader, Node], Any], Loader: None = ...
+) -> None: ...
+@overload
+def add_constructor(tag: _Str, constructor: Callable[[_Constructor, Node], Any], Loader: Type[_Constructor] = ...) -> None: ...
```
The second overload shouldn't have a default value. The first overload covers the case where the default value is used.

```suggestion
def add_constructor(tag: _Str, constructor: Callable[[_Constructor, Node], Any], Loader: Type[_Constructor]) -> None: ...
```

Stannislav replied:
Makes sense, thanks for pointing out. Fixed all four cases in dae51a6.

Akuli commented on stubs/PyYAML/yaml/__init__.pyi:
```diff
@@ -258,12 +261,46 @@ def safe_dump(
+) -> None: ...
+@overload
+def add_multi_constructor(
+    tag_prefix: _Str, multi_constructor: Callable[[_Constructor, _Str, Node], Any], Loader: Type[_Constructor] = ...
```
Same here:

```suggestion
    tag_prefix: _Str, multi_constructor: Callable[[_Constructor, _Str, Node], Any], Loader: Type[_Constructor]
```

Stannislav replied:
dae51a6

Akuli commented on stubs/PyYAML/yaml/__init__.pyi:
```diff
@@ -258,12 +261,46 @@ def safe_dump(
+def add_representer(data_type: Type[_T], representer: Callable[[Dumper, _T], Node]) -> None: ...
+@overload
+def add_representer(
+    data_type: Type[_T], representer: Callable[[_Representer, _T], Node], Dumper: Type[_Representer] = ...
```
Slightly different from the previous ones, but still, the second overload is used only when a `Dumper` argument is given.

```suggestion
    data_type: Type[_T], representer: Callable[[_Representer, _T], Node], Dumper: Type[_Representer]
```

Stannislav replied:
dae51a6

Akuli commented on stubs/PyYAML/yaml/__init__.pyi:
```diff
@@ -258,12 +261,46 @@ def safe_dump(
+def add_multi_representer(data_type: Type[_T], multi_representer: Callable[[Dumper, _T], Node]) -> None: ...
+@overload
+def add_multi_representer(
+    data_type: Type[_T], multi_representer: Callable[[_Representer, _T], Node], Dumper: Type[_Representer] = ...
```
Same here:

```suggestion
    data_type: Type[_T], multi_representer: Callable[[_Representer, _T], Node], Dumper: Type[_Representer]
```

Stannislav replied:
dae51a6

Akuli reviewed (approved):
Thanks!
